src/math/backends/gradients.ts

```diff
@@ -0,0 +1,216 @@
+ *
+ * @param f The function to execute. f() should return an NDArray of the same
+ * shape and dtype as dy.
+ * @param x The input to compute dy/dx over. This can be a single value or
+ * an object mapping a string to an NDArray. If using the object mode, this
+ * method will return an object of the same shape.
+ */
+export function vjp<T extends NDArray|NamedArrayMap, R extends Rank>(
+    f: () => NDArray<R>, x: T, dy: NDArray<R>): T {
+  const keys = x instanceof NDArray ? null : Object.keys(x);
+  const xs = util.flattenNameArrayMap(x, keys);
+
+  const gradientsMode = true;
+  const vjp = tidy('vjp', () => {
+    const y = f();
+    if (!util.arraysEqual(y.shape, dy.shape)) {
+      throw new Error(
+          `Cannot compute vector jacobian product, ` +
+          `y shape (${y.shape}) does not match dy shape (${dy.shape}).`);
+    }
+    return ENV.engine.gradientWrt(y, xs, dy);
+  }, gradientsMode);
+
+  if (x instanceof NDArray) {
+    return vjp[0] as T;
+  } else {
+    return util.unflattenToNameArrayMap(keys, vjp) as T;
+  }
+}
+
+/**
+ * Computes and returns the gradient of f(x) with respect to x.
+ *
+ * @param f The function to execute. f() should return a scalar.
+ *          TODO(nsthorat): Accept non-scalars.
+ * @param x The input to compute de/dx over. This can be a single value or
+ * an object mapping a string to an NDArray. If using the object mode, this
+ * method will return an object of the same shape.
+ */
+export function gradients<T extends NDArray|NamedArrayMap>(
+    f: () => Scalar, x: T): T {
+  const keys = x instanceof NDArray ? null : Object.keys(x);
+  const xs = util.flattenNameArrayMap(x, keys);
+
+  const returnValue = false;
+  const gradients = ENV.engine.gradients(f, xs, returnValue) as NDArray[];
+
+  if (x instanceof NDArray) {
+    return gradients[0] as T;
+  } else {
+    return util.unflattenToNameArrayMap(keys, gradients) as T;
+  }
+}
+
+/**
+ * Computes and returns the gradient of f(x) with respect to the list of
+ * trainable variables provided by `varList`. If no list is provided, it
+ * defaults to all trainable variables.
+ * @param f The function to execute. f() should return a scalar.
+ * @param varList An optional list of variables to provide gradients with
+ * respect to. Defaults to all trainable variables.
+ */
+export function variableGradients(f: () => Scalar, varList?: Variable[]):
+    {value: Scalar, gradients: NamedArrayMap} {
+  if (varList == null) {
+    // Get all of the trainable variables.
+    varList = [];
+    const varNames = Object.keys(ENV.engine.registeredVariables);
+    for (let i = 0; i < varNames.length; i++) {
+      const variable = ENV.engine.registeredVariables[varNames[i]];
+      if (variable.trainable) {
+        varList.push(variable);
+      }
+    }
+  } else {
+    // Prune non-trainable variables.
+    varList = varList.filter(variable => variable.trainable);
+  }
+
+  const gradientsMode = true;
```

Comment:
let's push here and below into the engine. this in-function input computation should be hidden in the sugar floating function